rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+
+    let msg = "HTTP protocol is not allowed for external requests, please use HTTPS.\n";
+    builder.body(String::from(msg)).unwrap()
+}
+
+// build a router for the HTTP -> HTTPS redirection
+// if the redirection URL cannot be built from the original request it returns error 400
+// instead of the redirection
+fn https_redirect() -> Router {
+    // see https://docs.rs/axum/latest/axum/routing/struct.Router.html#example
+    let redirect_service = tower::service_fn(|req: axum::extract::Request| async move {
+        let request_host = req.headers().get("host");
+        match request_host {
+            // missing host in the request header, we cannot build the redirection URL, return error 400
+            None => Ok(redirect_error()),
+            Some(host) => {
+                let host_string = host.to_str();
+                match host_string {
+                    // invalid host value in the request
+                    Err(_) => Ok(redirect_error()),
+                    Ok(host_str) => return Ok(redirect_https(host_str, req.uri())),
+                }
+            }
+        }
+    });
+
+    Router::new()
+        // the wildcard path below does not match an empty path, we need to match it explicitly
+        .route_service("/", redirect_service)
+        .route_service("/*path", redirect_service)
+}
+
+// start the web server
+async fn start_server(address: String, service: Router, ssl_acceptor: SslAcceptor) {
+    tracing::info!("Starting Agama web server at {}", address);
+
+    // see https://github.com/tokio-rs/axum/blob/main/examples/low-level-openssl/src/main.rs
+    // how to use axum with openSSL
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|_| panic!("could not listen on {}", &address));
+
+    pin_mut!(listener);
+
+    let redirector = https_redirect();
+
+    loop {
+        let tower_service = service.clone();
+        let redirector_service = redirector.clone();
+        let tls_acceptor = ssl_acceptor.clone();
+
+        // Wait for a new tcp connection
+        let (tcp_stream, addr) = listener.accept().await.unwrap();
+
+        tokio::spawn(async move {
+            if is_ssl_stream(&tcp_stream).await {
+                // handle HTTPS connection
+                let ssl = Ssl::new(tls_acceptor.context()).unwrap();
+                let mut tls_stream = SslStream::new(ssl, tcp_stream).unwrap();
+                if let Err(err) = SslStream::accept(Pin::new(&mut tls_stream)).await {
+                    tracing::error!("Error during TSL handshake from {}: {}", addr, err);
+                }
+
+                let stream = TokioIo::new(tls_stream);
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        tower_service.clone().call(request)
+                    });
+
+                let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
```

Comment:
This looks like too much qualification, to be honest 😅. Bear in mind that you can even rename the elements you import (see [The use declaration](https://doc.rust-lang.org/rust-by-example/mod/use.html)).